tests/static_map/heterogeneous_lookup_test.cu

```diff
@@ -0,0 +1,122 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <utils.hpp>
+
+#include <cuco/static_map.cuh>
+
+#include <thrust/device_vector.h>
+#include <thrust/execution_policy.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/iterator/transform_iterator.h>
+#include <thrust/transform.h>
+
+#include <catch2/catch.hpp>
+
+#include <tuple>
+
+// insert key type
+template <typename T>
+struct key_pair {
+  T a;
+  T b;
+
+  __host__ __device__ key_pair() {}
+  __host__ __device__ key_pair(T x) : a{x}, b{x} {}
+
+  // Device equality operator is mandatory due to libcudacxx bug:
+  // https://github.com/NVIDIA/libcudacxx/issues/223
+  __device__ bool operator==(key_pair const& other) const { return a == other.a and b == other.b; }
+};
+
+// probe key type
+template <typename T>
+struct key_triplet {
+  T a;
+  T b;
+  T c;
+
+  __host__ __device__ key_triplet() {}
+  __host__ __device__ key_triplet(T x) : a{x}, b{x}, c{x} {}
+
+  // Device equality operator is mandatory due to libcudacxx bug:
+  // https://github.com/NVIDIA/libcudacxx/issues/223
+  __device__ bool operator==(key_triplet const& other) const
+  {
+    return a == other.a and b == other.b and c == other.c;
+  }
+};
+
+// User-defined device hasher
+struct custom_hasher {
+  template <typename CustomKey>
+  __device__ uint32_t operator()(CustomKey const& k)
+  {
+    return thrust::raw_reference_cast(k).a;
+  };
+};
+
+// User-defined device key equality
+struct custom_key_equal {
+  template <typename LHS, typename RHS>
+  __device__ bool operator()(LHS const& lhs, RHS const& rhs)
+  {
+    return thrust::raw_reference_cast(lhs).a == thrust::raw_reference_cast(rhs).a;
+  }
+};
+
+TEMPLATE_TEST_CASE_SIG("User defined key and value type",
+                       "",
+                       ((typename Key, typename Value), Key, Value),
+#ifndef CUCO_NO_INDEPENDENT_THREADS  // Key type larger than 8B only supported for sm_70 and up
+                       (key_pair<int64_t>, int64_t),
+#endif
+                       (key_pair<int32_t>, int32_t))
+{
+  auto const sentinel_key   = Key{-1};
+  auto const sentinel_value = Value{-1};
+
+  constexpr std::size_t num      = 100;
+  constexpr std::size_t capacity = num * 2;
+  cuco::static_map<Key, Value> map{capacity,
+                                   cuco::sentinel::empty_key<Key>{sentinel_key},
+                                   cuco::sentinel::empty_value<Value>{sentinel_value}};
+
+  auto insert_pairs = thrust::make_transform_iterator(
+    thrust::counting_iterator<int>(0),
+    [] __device__(auto i) { return cuco::pair_type<Key, Value>(i, i); });
+  auto probe_keys = thrust::make_transform_iterator(
+    thrust::counting_iterator<int>(0), [] __device__(auto i) { return key_triplet<Value>(i); });
```

Comment:
Shouldn't `key_triplet` bet templated on `Key` not `Value`?

Reply:
The naming was a bit misleading. Corrected.